clang/lib/CodeGen/CGExprComplex.cpp

```diff
@@ -287,9 +288,47 @@ class ComplexExprEmitter
   ComplexPairTy EmitComplexBinOpLibCall(StringRef LibCallName,
                                         const BinOpInfo &Op);
 
-  QualType getPromotionType(QualType Ty) {
+  QualType GetHigherPrecisionFPType(QualType ElementType) {
+    const auto *CurrentBT = dyn_cast<BuiltinType>(ElementType);
+    switch (CurrentBT->getKind()) {
+    case BuiltinType::Kind::Float16:
+      return CGF.getContext().FloatTy;
+    case BuiltinType::Kind::Float:
+    case BuiltinType::Kind::BFloat16:
+      return CGF.getContext().DoubleTy;
+    case BuiltinType::Kind::Double:
+      return CGF.getContext().LongDoubleTy;
+    default:
+      return ElementType;
+    }
+  }
+
+  QualType HigherPrecisionTypeForComplexArithmetic(QualType ElementType,
+                                                   bool IsDivOpCode) {
+    QualType HigherElementType = GetHigherPrecisionFPType(ElementType);
+    const llvm::fltSemantics &ElementTypeSemantics =
+        CGF.getContext().getFloatTypeSemantics(ElementType);
+    const llvm::fltSemantics &HigherElementTypeSemantics =
+        CGF.getContext().getFloatTypeSemantics(HigherElementType);
+    if (llvm::APFloat::semanticsMaxExponent(ElementTypeSemantics) * 2 + 1 <=
+        llvm::APFloat::semanticsMaxExponent(HigherElementTypeSemantics)) {
+      return CGF.getContext().getComplexType(HigherElementType);
+    } else {
+      FPHasBeenPromoted = LangOptions::ComplexRangeKind::CX_Improved;
```

Comment:
If I'm following correctly, this means that the behavior wouldn't be right for this code:

```c
_Complex float func(_Complex float a, _Complex long double b, _Complex float c) {
  return (_Complex float)(b / c) / a;
}
```

The code would first process the long double complex division, which isn't promotable, and then the float division, using the same `ComplexExprEmitter` I think, so the resetting to improved would cause the second float division to be emitted as improved instead of promoted.

Reply:
Yes I see the problem. Working on this. May be the` FPHasBeenPromoted` should be "attached" to the `BinOpInfo` instead?

Reply:
I do see 2 different `ComplexExprEmitter` each with a different value of `FPHasBeenPromoted` though.

Reply:
Can you add a test for this case?

Reply:
The test is at line #2319 of CodeGen/cx-complex-range.c.